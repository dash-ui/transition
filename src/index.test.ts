import styles from '@-ui/styles'
import transition from './index'

afterEach(() => {
  styles.dash.sheet.flush()
  document.getElementsByTagName('html')[0].innerHTML = ''
})

describe('transition()', () => {
  it('should write styles to the DOM', () => {
    const fade = transition({
      in: {
        opacity: 0,
        duration: 300,
      },
    })

    fade('in')
    expect(document.querySelectorAll(`style[data-dash]`).length).toBe(1)
    expect(document.querySelectorAll(`style[data-dash]`)[0]).toMatchSnapshot()
  })

  it('should return class name', () => {
    const fade = transition({
      in: {
        opacity: 0,
        duration: 300,
      },
    })

    expect(fade('in')).toMatchSnapshot()
  })

  it('should override default duration', () => {
    const fade = transition({
      default: {
        duration: 100,
      },
      in: {
        opacity: 0,
        duration: 300,
      },
    })

    expect(fade.css('in')).toMatchSnapshot()
  })

  it('should override default timing', () => {
    const fade = transition({
      default: {
        timing: 'linear',
        duration: 100,
      },
      in: {
        opacity: 0,
        timing: 'ease-in',
      },
    })

    expect(fade.css('in')).toMatchSnapshot()
  })

  it('should override default delay', () => {
    const fade = transition({
      default: {
        delay: 100,
        duration: 100,
      },
      in: {
        opacity: 0,
        delay: 300,
      },
    })

    expect(fade.css('in')).toMatchSnapshot()
  })

  it('should have special handling for transforms', () => {
    const transforms = {
      matrix: [1, 1],
      matrix3d: [1, 1, 1],
      perspective: 1,
      rotate: 1,
      rotateX: 1,
      rotateY: 1,
      rotateZ: 1,
      scale: [1, 1],
      scaleX: 1,
      scaleY: 1,
      scaleZ: 1,
      scale3d: [1, 1, 1],
      skew: [1, 1],
      skewX: 1,
      skewY: 1,
      translate3d: [1, 1, 1],
      translate: [1, 1],
      x: 1,
      y: 1,
      z: 1,
    }

    for (const [key, value] of Object.entries(transforms)) {
      const t = transition({
        default: {
          duration: 100,
        },
        in: {
          [key]: value,
        },
      })

      expect(t.css('in')).toMatchSnapshot()
    }
  })

  it('should concat multiple transforms', () => {
    const t = transition({
      default: {
        duration: 100,
      },
      in: {
        x: 12,
        y: 13,
        matrix: [1, 1],
      },
    })

    expect(t.css('in')).toMatchSnapshot()
  })

  it('should concat multiple transitions', () => {
    const t = transition({
      default: {
        duration: 100,
      },
      slide: {
        x: 12,
        y: 13,
        matrix: [1, 1],
      },
      fadeIn: {
        opacity: 1,
        duration: 50,
      },
      fadeOut: {
        opacity: 0,
        duration: 30,
      },
    })

    expect(t.css('slide', 'fadeOut', 'fadeIn')).toMatchSnapshot()
  })

  it('should create transitions from object arguments', () => {
    const t = transition({
      default: {
        duration: 100,
      },
      slide: {
        x: 12,
        y: 13,
        matrix: [1, 1],
      },
      fadeIn: {
        opacity: 1,
        duration: 50,
      },
      fadeOut: {
        opacity: 0,
        duration: 30,
      },
    })

    expect(
      t.css({slide: true, fadeIn: true}, {fadeOut: false})
    ).toMatchSnapshot()
  })

  it('should compose transitions', () => {
    const slow = transition({
      default: {
        duration: 1000,
      },
    })

    const t = transition(slow, {
      slide: {
        x: 12,
        y: 13,
        matrix: [1, 1],
      },
      fadeIn: {
        opacity: 1,
      },
      fadeOut: {
        opacity: 0,
      },
    })

    expect(
      t.css({slide: true, fadeIn: true}, {fadeOut: false})
    ).toMatchSnapshot()
  })

  it('should compose transitions to style object', () => {
    const t = transition({
      default: {
        duration: 1000,
      },
      slide: {
        x: 12,
        y: 13,
        matrix: [1, 1],
        duration: 300,
      },
      fadeIn: {
        opacity: 1,
      },
      fadeOut: {
        opacity: 0,
      },
    })

    expect(
      t.style({slide: true, fadeIn: true}, {fadeOut: false})
    ).toMatchSnapshot()
  })

  it('should create new instance', () => {
    const newStyles = styles.create()
    const t = transition.create(newStyles)
    expect(t.dash).toBe(newStyles.dash)
  })
})
