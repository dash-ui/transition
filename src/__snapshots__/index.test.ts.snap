// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transition() should compose transitions 1`] = `"transform:translateX(12px) translateY(13px) matrix(1,1);opacity:1;transition:transform 1000ms,opacity 1000ms;"`;

exports[`transition() should compose transitions to style object 1`] = `
Object {
  "opacity": 1,
  "transform": "translateX(12px) translateY(13px) matrix(1,1)",
  "transition": "transform 300ms,opacity 1000ms",
}
`;

exports[`transition() should concat multiple transforms 1`] = `"transform:translateX(12px) translateY(13px) matrix(1,1);transition:transform 100ms;"`;

exports[`transition() should concat multiple transitions 1`] = `"transform:translateX(12px) translateY(13px) matrix(1,1);opacity:1;transition:transform 100ms,opacity 30ms,opacity 50ms;"`;

exports[`transition() should create default phase with a function 1`] = `
Object {
  "transform": "translateX(12px) translateY(13px) matrix(1,1)",
  "transition": "transform var(--duration-slow)",
}
`;

exports[`transition() should create phase with a function 1`] = `
Object {
  "transform": "translateX(12px) translateY(13px) matrix(1,1)",
  "transition": "transform var(--duration-slow)",
}
`;

exports[`transition() should create transitions from object arguments 1`] = `"transform:translateX(12px) translateY(13px) matrix(1,1);opacity:1;transition:transform 100ms,opacity 50ms;"`;

exports[`transition() should have special handling for transforms 1`] = `"transform:matrix(1,1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 2`] = `"transform:matrix3d(1,1,1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 3`] = `"transform:perspective(1px);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 4`] = `"transform:rotate(1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 5`] = `"transform:rotateX(1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 6`] = `"transform:rotateY(1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 7`] = `"transform:rotateZ(1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 8`] = `"transform:scale(1,1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 9`] = `"transform:scaleX(1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 10`] = `"transform:scaleY(1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 11`] = `"transform:scaleZ(1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 12`] = `"transform:scale3d(1,1,1);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 13`] = `"transform:skew(1deg,1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 14`] = `"transform:skewX(1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 15`] = `"transform:skewY(1deg);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 16`] = `"transform:translate3d(1px,1px,1px);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 17`] = `"transform:translate(1px,1px);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 18`] = `"transform:translateX(1px);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 19`] = `"transform:translateY(1px);transition:transform 100ms;"`;

exports[`transition() should have special handling for transforms 20`] = `"transform:translateZ(1px);transition:transform 100ms;"`;

exports[`transition() should ignore null values 1`] = `"transform:translateY(0px);transition:transform 100ms 300ms;"`;

exports[`transition() should override default delay 1`] = `"opacity:0;transition:opacity 100ms 300ms;"`;

exports[`transition() should override default duration 1`] = `"opacity:0;transition:opacity 300ms;"`;

exports[`transition() should override default timing 1`] = `"opacity:0;transition:opacity 100ms ease-in;"`;

exports[`transition() should return class name 1`] = `"ui-zi2gip"`;

exports[`transition() should write styles to the DOM 1`] = `
<style
  data-dash="ui"
>
  .ui-zi2gip{opacity:0;-webkit-transition:opacity 300ms;transition:opacity 300ms;}
</style>
`;
